### PR TITLE
Updates to saving true particle tracks

### PR DESCRIPTION
**Pull request by nickwp**
This changes how true particle tracks are chosen to be saved.

Previously, this logic was spread in two places: the EndOfEventAction and the TrackingAction. The TrackingAction marks events to be saved, while EndOfEventAction chooses some of these to save (while flat tree format saves all marked to be saved).

Now, all tracks marked to be saved by TrackingAction are saved. Optionally, it also saves tracks if they are in the chain of particles from primary particles until (but not including) photons that produce Cherenkov hits. (Cherenkov hit producing photons have their details already saved in the WCSimRootCherenkovHitTime class - later this could be removed and just save the track, but for now avoiding breaking any analysis code relying on that).

The option to save all tracks involved in producing hits can be turned on or off with /Tracking/saveHitProducingTracks

The logic for marking tracks to be saved is also simplified:
Anything in the ParticleList
Anything in the ProcessList
These lists are left the same as they were before, but can be added to by mac file (no change).

Here are benchmarks for 100 x 500 MeV electrons with the new version (with and without the saveHitProducingTracks) and old version:
**saveHitProducingTracks true**
        User time (seconds): 148.27
        Maximum resident set size (kbytes): 204,528
        File size (kbytes): 39,071
**saveHitProducingTracks false**
        User time (seconds): 147.47
        Maximum resident set size (kbytes): 204,500
        File size (kbytes): 38,917
**old version**
        User time (seconds): 151.38
        Maximum resident set size (kbytes): 204,488
        File size (kbytes): 38,197

So only about 2% increase in file size, negligible change to memory usage and processing time.

### CODE REVIEW

File: src/WCSimWCSD.cc
Comment (kmtsui) [inline src/WCSimWCSD.cc]: In the original code, this will be the most upstream parent ID. Now it becomes the current track ID. How can we get the parent ID in the new code?

Comment (tdealtry) [reply src/WCSimWCSD.cc]: Since this is meant as a seemless low-impact change, I'd recommend keeping the primary parent ID definiton the same as it currently is. But add the extra information of parent ID to everything through to and including WCSimRootCherenkovHit

Comment (nickwp) [reply src/WCSimWCSD.cc]: Ah, I'm sorry I forgot to also mention this 'fix'.

Actually, the current version does NOT save the primary parent in this variable. The track ID it does save is somewhat complicated... In the new verison it will take the direct parent if that parent track is saved. In the old version, it will be the same if the direct parent is a saved track and is anything other than a gamma (set [here](https://github.com/WCTE/WCSim/blob/32640318f1ef19fad8b15669242c4d63f6da65dd/src/WCSimTrackingAction.cc#L182-L186)), or if it is a gamma produced by a pi0 or nCapture or NeutronInelastic (set [here](https://github.com/WCTE/WCSim/blob/32640318f1ef19fad8b15669242c4d63f6da65dd/src/WCSimTrackingAction.cc#L180)).

So there are only two cases where anything is changed:

1. The choice of which true tracks are saved as WCSimRootTrack has been changed by the rest of this PR, and this can change whether that track would be the parent or not (because it never saves a parent ID of a track that isn't saved. This will change even if we kept the old code for the hit parent ID, since both old and new require the track to be saved.
2.  If the particle that produces a hit is itself produced by a gamma that is not produced by a pii0 or nCapture or NeutronInelastic, then the old code would save the ID of the gamma's parent instead of the gamme. The new code saves the ID of the gamma. I can't think of any situation where a gamma directly produces an optical photon that produces a hit, so this situation should never occur anyway.

The original reason (I believe, based on the comments) for the old version's logic is that certain electron tracks in an EM shower were not saved before, so the code was to be careful that it didn't save a parent ID of a gamma in the EM shower that was saved (unless it was one of these specific ones) but instead always traced back to a particle capable of producing Cherenkov light, even if that wasn't the particle that produced the light (because that particle's track wasn't saved.

So in the end, the only way to retain the old version's functionality would be to add back in all the arbitrary logic about which tracks to save and just use that to set the parent ID of the hit, which doesn't seem useful since we're now saving the other tracks anyway.

Comment (tdealtry) [reply src/WCSimWCSD.cc]: Thanks for clearing up this complicated logic! I'm happy for the new code to do as you say

Just a request to make the variables/functions more descriptive of reality. So instead of "primary parent ID" or similar, something instead like "deepest saved track ID" (better options are potentially available. "Lowest" was my first thought, but doesn't work (the primary particles use the lowest numbers))

Comment (nickwp) [reply src/WCSimWCSD.cc]: Sure, I can do this later, about to board a flight... I had thought about renaming but just left it as it was, but agree it's better to change. Actually as it gets passed throughout the code, its variously called "primaryParentId" or "primParentId" or just "parentId". In the WCSimRootCherenkovHitTime class that users actually see, I think the method is simply "GetParentID".  I'd rather not change this (to avoid breaking anything downstream), so maybe to match that we just call it parentID in the internal variables too? Or if you want it to be very explicit, I'd go for something like savedParentTrackID or parentSavedTrackID since the one requirement is that the track is saved. What do you think?

Comment (tdealtry) [reply src/WCSimWCSD.cc]: > savedParentTrackID or parentSavedTrackID

Yeah these are good. I like both

And I do think being explicit rather than adherring to old unclear convetions is better.
Since the only thing 95% of users will see is the method in `WCSimRootCherenkovHitTime`, then I think making the rest of the code use the new name would be good. And inside `WCSimRootCherenkovHitTime` maybe deprecate `GetParentID()` whilst adding e.g. `GetParentSavedTrackID()`

Enjoy your travels!

**Comment by tdealtry** [comment]
Just to say I've looked through the rest of the PR. All looks great. Thanks for doing this important job @nickwp!

File: src/WCSimWCSD.cc
Comment (nickwp) [reply src/WCSimWCSD.cc]: I've made these changes to variable names now, and also created the new GetParentSavedTrackID() method to depricate the existing GetPerentID().

**Comment by nickwp** [comment]
I'm waiting for some additional checks on whether after these changes it's now possible to choose to save optical photon tracks (even those that don't produce hits) by specifying in the particle list or process list. This was a requested feature that previously had many hard-coded places blocking that from being possible, which should be removed now.
But I think there might need to be additional small changes for that to work, so we can either wait to include that in this PR before merging, or merge this first and then make the additional fix later, if that is preferred.

**Comment by nickwp** [comment]
I made some additional updates to ensure that:

* Optical photon tracks can be saved if explcitly enabled by `/Tracking/trackParticle 0` (because Geant4 gives optical photons PDG code of 0, rather than 22 used for other gammas) or e.g. `/Tracking/trackProcess Cerenkov` to save tracks produced by the Cherenkov process. Previously there was code always preventing optical photons being tracked.
* Optical photon tracks are not saved by default, even if they produce a hit, since there's a large number of them and the hit-producing photon details are already saved in WCSimRootCherenkovHitTime. The tracks of the parents of optical photons producing hits are still saved.
* The existing macro command `/Tracking/fractionOpticalPhotonsToDraw [percent value]` still controls how many optical photon processes are tracked, by drawing a random number from 0 to 1 at the start of processing any optical photon and if that random number is below the chosen fraction then it gets tracked. So this can still be used to choose what percentage of optical photons to actually track and save. Previously, this command was available for tracking optical photons but the actual saving of optical photon tracks to the output was blocked by other code.

**Comment by nickwp** [comment]
I think this is ready to merge, if someone is able to do a final review?

Repeated benchmarks with the updated version, just to be sure no problems introduced:
 **saveHitProducingTracks true**
User time (seconds): 153.71
Maximum resident set size (kbytes): 155,280
File size (kbytes): 39,072
**saveHitProducingTracks false**
User time (seconds): 156.01
Maximum resident set size (kbytes): 155,228
File size (kbytes): 38,918
**old version**
User time (seconds): 151.38
Maximum resident set size (kbytes): 204,488
File size (kbytes): 38,197

**Review by tdealtry — CHANGES_REQUESTED** [review]
Thanks @nickwp!

My main comment is why is this PR not on WCTE/WCSim? @kmtsui I thought we agreed WCSim/WCSim was going to be used for all (HKFD/IWCD/WCTE) development? Are we still transitioning towards that?

One minor comment is to please _document_ the new mac file options in WCSim.mac

**Comment by nickwp** [comment]
> My main comment is why is this PR not on WCTE/WCSim? @kmtsui I thought we agreed WCSim/WCSim was going to be used for all (HKFD/IWCD/WCTE) development? Are we still transitioning towards that?

Ka Ming and I discussed this before making the PR and at the time I think we decided to try to more quickly (and temporarily) get this change into this WCSim/develop branch of WCTE/WCSim (which is the branch Ka Ming kept in sync with WCSim/WCSim), before a second PR with full review into the develop branch of WCSim/WCSim... In the end, the review happened here anyway, so if this is merged I can immediately PR into develop branch of WCSim/WCSim. In future I think we should just go directly to develop branch of WCSim/WCSim.

> One minor comment is to please document the new mac file options in WCSim.mac

I've added to WCSim.mac and documented the new command added by this PR: `/Tracking/saveHitProducingTracks`
Also took the opportunity to do the same for the pre-existing but currently undocumented commands: `/Tracking/trackParticle`, `/Tracking/trackProcess`, and `/Tracking/fractionOpticalPhotonsToDraw` (which was already included in WCSim.mac but was not documented)

**Comment by tdealtry** [comment]
This is great thanks @nickwp I'm happy for you to merge here (then PR to WCSim/WCSim) whenever you're ready

**Pull request merged**
